chain-crypto/src/digest.rs

```diff
@@ -242,3 +246,80 @@ impl<H: DigestAlg> fmt::Debug for Digest<H> {
         f.write_str(")")
     }
 }
+
+impl<H: DigestAlg> Digest<H> {
+    /// Get the digest of a slice of data
+    pub fn digest(slice: &[u8]) -> Self {
+        let mut ctx = Context::new();
+        ctx.append_data(slice);
+        ctx.finalize()
+    }
+}
+
+use std::marker::PhantomData;
+
+/// A typed version of Digest
+pub struct DigestOf<H: DigestAlg, T> {
+    inner: Digest<H>,
+    marker: PhantomData<T>,
+}
+
+impl<H: DigestAlg, T> Clone for DigestOf<H, T> {
+    fn clone(&self) -> Self {
+        DigestOf {
+            inner: self.inner.clone(),
+            marker: self.marker.clone(),
+        }
+    }
+}
+
+impl<H: DigestAlg, T> From<DigestOf<H, T>> for Digest<H> {
+    fn from(d: DigestOf<H, T>) -> Self {
+        d.inner
+    }
+}
+
+impl<H: DigestAlg, T> From<Digest<H>> for DigestOf<H, T> {
+    fn from(d: Digest<H>) -> Self {
+        DigestOf {
+            inner: d,
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<H: DigestAlg, T> DigestOf<H, T> {
+    /// Coerce a digest of T, to a digest of U
+    pub fn coerce<U>(&self) -> DigestOf<H, U> {
```

Comment:
This also looks odd. Should any untyped comparisons and the like be done on untyped `Digest`?

Reply:
this is an escape hatch when type-level situation is not ideal, just like `unsafe` is useful sometimes. This is not about untyped comparaison, it's about making it quack like a duck in some, hopefully rare, situation.

Reply:
How does this look odd? You may compute the `DigestOf` an array of bytes (i.e. `Vec<u8>`) and now you want to lift the tag to `Fragment` as `Vec<u8>` was only the serialised version of `Fragment`.

This is useful to have an explicit conversion from one type to another as we grep `coerce` and see where the type tag has been changed to another. Allowing `From<Digest<T, H>> for Digest<T, U>` would have been a bad idea, at least the unsafety is explicit in the sense it is not a standard API feature that can be silently applied.

Reply:
This sounds scary in conjunction with cryptographic hashes and the notion of a type that looks to a code reader like "this is _the_ cryptographic digest of a T".

Reply:
@NicolasDP the `coerce` might be hidden in a totally different part of the code.